### PR TITLE
op-core: add OP tx JSON codecs and consensus receipt encoding

### PR DESCRIPTION
**Pull request by sebastianst**
Part of #20264 (opgeth-decoupling: make op-service/sources the OP-aware eth client). First of three stacked PRs; the follow-ups make `op-service/sources` decode L2 blocks/receipts via these codecs and migrate op-batcher to payload-based block loading.

Adds to `op-core/types`:

- **`DepositTx`/`PostExecTx` JSON codecs** matching op-geth's RPC transaction wire format (validation arms mirrored field-by-field), incl. accepting-and-discarding the post-Regolith effective deposit nonce. `PostExecTx.Hash` follows the canonical EIP-2718 rule, `keccak256(0x7D ‖ Data)` — shared with op-reth and pinned by a cross-client golden vector (cross-client alignment landed via ethereum-optimism/op-geth#805 + #21931).
- **`Receipts`**, a `DerivableList` for receipts-root recomputation with the three OP arms hand-rolled: post-Canyon deposit receipts (nonce+version in the hash preimage), pre-Canyon deposit receipts (nonce deliberately excluded from the root derivation — diverges from `MarshalBinary`, per op-geth), and `0x7D` post-exec receipts.
- **`Receipt` wrapper methods** — OP-aware `MarshalBinary`/`UnmarshalBinary` (the `debug_getRawReceipts` path), `EncodeRLP`/`DecodeRLP`, `DeriveFields`, and `Size`, overriding every promoted go-ethereum method that touches OP state so deposit receipts encode/derive from the authoritative outer `DepositNonce`/`DepositReceiptVersion` fields and `0x7E`/`0x7D` decode survives the upstream cutover.
- **`Receipts.Geth`**, a non-copying view of the embedded upstream receipts.

Everything is pinned byte-for-byte against op-geth by differential tests (removed at the final cutover); the deposit hash-preimage rules and codec round-trips additionally get cutover-surviving tests.

🤖 *Co-created with Claude Fable 5*

### CODE REVIEW

File: op-core/types/transaction_json.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	d.IsSystemTransaction = dec.IsSystemTx != nil && *dec.IsSystemTx
+	return nil
+}
+
+// Hash returns the transaction hash as op-geth computes it:
+// keccak256(PostExecTxType || RLP([Data])). Note this is NOT the hash of the
+// canonical envelope — the wire encoding appends Data verbatim (see
+// MarshalBinary), but op-geth hashes typed transactions over the RLP of the
+// inner struct, and for PostExecTx the two differ.
+func (p *PostExecTx) Hash() common.Hash {
```
Comment (joshklop) [inline op-core/types/transaction_json.go:134]: **[P1] Resolve the post-exec hash mismatch with op-reth.** This computes `keccak256(0x7d || RLP([Data]))`, matching op-geth, but `TxPostExec::tx_hash` in `rust/op-alloy/crates/consensus/src/post_exec.rs` hashes its canonical EIP-2718 encoding, `keccak256(0x7d || Data)`. The stacked sources change uses this hash to fetch receipts, so a Lagoon block served by op-reth can request the trailing post-exec receipt under the wrong hash. Please resolve the cross-client behavior and add an op-reth/op-alloy parity vector rather than pinning only to op-geth.

File: op-core/types/receipt_consensus.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		types.Receipts{&r.Receipt}.EncodeIndex(0, w)
+	}
+}
+
+// UnmarshalBinary decodes the consensus encoding of a receipt, routing the
+// OP Stack synthetic receipt types (0x7E deposit, 0x7D post-exec) to this
+// package and everything else to go-ethereum. The OP JSON-only fee fields
+// (L1GasPrice, OperatorFeeScalar, ...) are not part of the consensus encoding
+// and remain nil.
+func (r *Receipt) UnmarshalBinary(b []byte) error {
```
Comment (joshklop) [inline op-core/types/receipt_consensus.go:223]: **[P2] Override the promoted receipt binary encoder.** `Receipt` now has an OP-aware `UnmarshalBinary`, but `MarshalBinary` is still promoted from the embedded geth receipt. After the upstream cutover that encoder cannot include deposit nonce/version; even before cutover it reads the embedded shadow fields rather than the authoritative outer fields on a manually wrapped receipt. Thus an OP deposit receipt cannot reliably binary-round-trip. Please add a local `MarshalBinary` that encodes deposit receipts from the outer `DepositNonce`/`DepositReceiptVersion` fields.

**Review by joshklop — COMMENTED** [review]
Two correctness issues to address before this stack lands.

File: op-core/types/transaction_json.go (hunk 1 above)
Comment (sebastianst) [reply op-core/types/transaction_json.go:134]: 🤖 *Claude says:* Resolved cross-client rather than pinned to op-geth: op-geth's `PostExecTx.Hash` is fixed to the canonical EIP-2718 rule in ethereum-optimism/op-geth#805, and #21931 bumps the pin, adds the op-alloy golden vector, and asserts the op-reth-served hash in the sdm acceptance tests. This stack is rebased onto #21930: `optypes.PostExecTx.Hash` now hashes the canonical encoding (covered by the shared cross-client golden vector), and the sources hash path lost its 0x7D special case entirely — `keccak(raw)` is now correct for every tx class.

File: op-core/types/receipt_consensus.go (hunk 2 above)
Comment (sebastianst) [reply op-core/types/receipt_consensus.go:223]: 🤖 *Claude says:* Fixed — `optypes.Receipt` now defines an OP-aware `MarshalBinary` that encodes deposit receipts from the authoritative outer `DepositNonce`/`DepositReceiptVersion` fields (and handles `0x7D`), delegating standard types to go-ethereum. Added `TestReceiptBinaryRoundTrip` (op-geth-independent, survives the cutover) and `TestReceiptMarshalBinaryDifferential` vs op-geth. Both fixes land in the `op-core: harden OP tx/receipt codecs after review` commit on this branch.

File: op-core/types/transaction_json.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+// effective deposit nonce that mined deposits carry on the RPC wire is not
+// part of this type (or of the canonical encoding), so a decode/encode
+// round-trip does not preserve it.
+func (d *DepositTx) MarshalJSON() ([]byte, error) {
+	typ := hexutil.Uint64(DepositTxType)
+	enc := &txJSON{
+		Type:       &typ,
+		To:         d.To,
+		Gas:        (*hexutil.Uint64)(&d.Gas),
+		Value:      (*hexutil.Big)(d.Value),
```
Comment (joshklop) [inline op-core/types/transaction_json.go]: **[P2] Normalize nil deposit values when marshaling JSON.** `DepositTx.MarshalBinary` supports a nil `Value`, and the existing test cases treat it as valid. However, this emits `"value": null`, which this type's own `UnmarshalJSON` rejects. Op-geth normalizes nil to zero and emits `"0x0"`. Please normalize nil to `new(big.Int)` and add a direct JSON round-trip test for the nil-value case.

File: op-core/types/receipt_consensus.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	}
+}
+
+// MarshalBinary returns the consensus encoding of the receipt, routing the
+// OP Stack synthetic receipt types (0x7E deposit, 0x7D post-exec) to this
+// package and everything else to go-ethereum. The deposit encoding reads the
+// authoritative OUTER DepositNonce/DepositReceiptVersion fields and — unlike
+// the receipts-root derivation in EncodeIndex — always includes the nonce when
+// set, matching op-geth's MarshalBinary. Inverse of UnmarshalBinary.
+func (r *Receipt) MarshalBinary() ([]byte, error) {
```
Comment (joshklop) [inline op-core/types/receipt_consensus.go:202]: **[P2] Override the promoted receipt RLP codec.** Adding `MarshalBinary`/`UnmarshalBinary` does not replace the embedded geth receipt's promoted `EncodeRLP`/`DecodeRLP`. Consequently, `rlp.Encode` reads the embedded shadow fields and omits the authoritative outer deposit nonce/version fields. `rlp.Decode` populates only the shadows today and will reject `0x7D`/`0x7E` after switching to upstream geth. Please add OP-aware `EncodeRLP` and `DecodeRLP` methods with a deposit-receipt RLP round-trip test.

**Review by joshklop — COMMENTED** [review]
Non-blocking: the PR description still says post-exec hashes use `0x7D || RLP([Data])`; the implementation now correctly hashes `0x7D || Data`.

**Review by joshklop — APPROVED** [review]
Optimistically stamping for now.

File: op-core/types/transaction_json.go (hunk 3 above)
Comment (sebastianst) [reply op-core/types/transaction_json.go]: 🤖 *Claude says:* Fixed in `op-core: override promoted receipt RLP codec, normalize nil deposit value` — `MarshalJSON` normalizes a nil `Value` to zero like op-geth, and `TestDepositTxJSONNilValueRoundTrip` pins the direct round-trip (plus binary-encoding agreement between the nil and decoded-zero forms).

File: op-core/types/receipt_consensus.go (hunk 4 above)
Comment (sebastianst) [reply op-core/types/receipt_consensus.go:202]: 🤖 *Claude says:* Fixed in the same commit — `Receipt` now overrides `EncodeRLP`/`DecodeRLP`: typed OP receipts encode as an RLP byte string wrapping the envelope from the OP-aware `MarshalBinary` (outer nonce/version authoritative), and the string-kind decode routes through `UnmarshalBinary`, so `0x7E`/`0x7D` survive the cutover. Covered by `TestReceiptRLPDifferential` (vs op-geth's `rlp.EncodeToBytes`) and the cutover-surviving `TestReceiptRLPRoundTrip`. You're right that this was the same promoted-member pattern as `MarshalBinary` — we've queued a full promoted-codec audit checklist (and a possible lint) as the durable fix.
